### ISSUE TITLE
CRI and Kubelet are way too docker specific

### ISSUE THREAD

**Issue by runcom** [kubernetes/kubernetes#47288]
The `CreateContainer` and `StartContainer` API in the CRI are too docker specific. The CRI isn't actually defining what `CreateContainer` means for a runtime and docker and containerd-shim behaves incorrectly.

In CRI-O when we get a `CreateContainer` CRI call, we run `runc create ...` - which I'm super sure is the correct way of handing a `CreateContainer` request. `runc create ...` can fail if, for instance, the command isn't in `$PATH`. That failure is a `CreateError` (!!!) but today it's reported as a `StartError` (!!!!)

However, Kubelet, containerd-cri, dockershim aren't actually handling `ContainerCreate` errors. Instead, because of docker, `docker create` never fails if command isn't in `$PATH` (test it yourself). That's because nor docker nor containerd actually call `runc create ...` when creating a container.
They call `runc create ...` as part of the `Start...` processes instead.

For Kubelet, containerd-cri and dockershim, `CreateContainer` means: "store a created container in memory in the runtime". This is totally wrong, or it must be clearly stated in the CRI API (but I'd argue create should actually _create_ a container).
This behavior is explained in how containerd-shim is handing `CreateContainer` here: https://github.com/kubernetes-incubator/cri-containerd/blob/master/pkg/server/container_create.go
There's no point in that code which actually `create` a container through containerd so that code can't fail if  command isn't in `$PATH`.

This flow is clearly a leftover from `docker create`/`docker start` flow and I strongly believe the CRI and the kubelet are too docker specific.

For instance, Kubelet expect a `StartError` (https://github.com/kubernetes-incubator/cri-containerd/blob/227dbe97f290e7fc7fc684f1be39cdcdd81fcc65/pkg/server/helpers.go#L47) but that is clearly not a `StartError` if it happens at container creation time. The `ContainerStart` function in containerd-shim is actually creating the container here https://github.com/kubernetes-incubator/cri-containerd/blob/master/pkg/server/container_start.go#L200 - if that fails it's not a start error (!!!) https://github.com/kubernetes-incubator/cri-containerd/blob/master/pkg/server/container_start.go#L87-L96 

I propose a full review of the interaction between the CRI and Kubelet wrt to container runtimes. We can't just assume the flow is the docker flow cause it isn't anymore.
We ran into this bug in CRI-O but since there's no clear documentation on how to handle Create/Start request, this must be revisited.

It's also super-semantically incorrect to return a `StartError` where it's actually a creation error. 

This create/start flow is just an example. I'm sure other APIs must be revisited.

@Random-Liu @mrunalp @feiskyer  PTAL

**Comment by runcom** [comment]
@kubernetes/sig-node-api-reviews

**Comment by runcom** [comment]
One other issue is, for instance, oomkill status from container

**Comment by cyphar** [comment]
@runcom I have to check the spec but I feel as though the OCI runtime-spec actually guarantees the opposite behaviour:

> runc create ... can fail if, for instance, the command isn't in $PATH

I remember discussions on this topic (effectively some users wanted to create the `$PATH` binary using hooks, or didn't even want to call `start` on a created container). We (runc) should double-check what the spec says though.

**Comment by feiskyer** [comment]
Kubelet doesn't depend on `StartError` or `Completed`, but it should indeed add more accurate descriptions on the reason of container state, e.g. OOMKilled.

We should make an agreement on the reason list first. @runcom What's your suggestion about this?

**Comment by runcom** [comment]
@feiskyer I'll put something together on Monday 

Also, I still think relies somehow on the wrong create/start behavior of Docker. For instance, in CRI-O we return error in create whether in Docker we do not. Not a kube expert so I'm just saying to check if erroring out on create like cri-o does is still correct for the kubelet.

**Comment by runcom** [comment]
> Kubelet doesn't depend on StartError or Completed, but it should indeed add more accurate descriptions on the reason of container state, e.g. OOMKilled.

so the container status reason is just a string described in CRI as https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/apis/cri/v1alpha1/runtime/api.proto#L757:

> Brief CamelCase string explaining why container is in its current state.

I propose to standardize the OOM killed status for a container as kubelet relies on it, PR here:https://github.com/kubernetes/kubernetes/pull/47306

**Comment by yujuhong** [comment]
> This flow is clearly a leftover from docker create/docker start flow and I strongly believe the CRI and the kubelet are too docker specific.

Yes, there is no denying that kubelet/CRI was heavily influenced by docker.  We welcome suggestions to improve all aspects of the API, but please keep in mind that the the docker integration is still the default and we have to continue support that for an indefinite time (unless we want to keep two implementations of kubelet). It is a compromise, and maybe we can retire that integration (with cri-containerd as a replacement?) sometime, but before that it's a technical debt we need to carry for a while. Anything that requires drastically changing kubelet and is completely incompatible with docker may have to wait for a while IMO. 

A lot of the points you mentioned originate from the different definition of container lifecycle in OCI and in CRI, which is closer to docker. I am as sure as you that one of them is clearly the *right* one. (If it was so clear, we could've just adopted OCI directly). I think there's room for discussion and I do think CRI lacks clear documentation. We will be working on that and help is appreciated.

> I propose to standardize the OOM killed status for a container as kubelet relies on it, PR here:#47306

That was wrong. I sent #47357 to fix the logic.



@runcom, thanks for filing the issue and please continue doing so! I'd like to more feedback and suggestions on the API. Even if some of them are not actionable right now, we can document and make sure we come back to it in the future :-)

**Comment by runcom** [comment]
> Yes, there is no denying that kubelet/CRI was heavily influenced by docker. We welcome suggestions to improve all aspects of the API, but please keep in mind that the the docker integration is still the default and we have to continue support that for an indefinite time (unless we want to keep two implementations of kubelet). It is a compromise, and maybe we can retire that integration (with cri-containerd as a replacement?) sometime, but before that it's a technical debt we need to carry for a while. Anything that requires drastically changing kubelet and is completely incompatible with docker may have to wait for a while IMO.

Right this is clear to me but thanks for the big disclaimer :) (s/containerd-cri/cri-o/ also? :)) 

About the documentation about CRI yes I'll keep adding stuff as I've already done in the past! 

@yujuhong let's leave this open for further discussion

**Comment by yujuhong** [comment]
> Right this is clear to me but thanks for the big disclaimer :) (s/containerd-cri/cri-o/ also? :))

Haha. I think cri-o is an alternative to docker, but cri-containerd could actually be a replacement in the sense that it could still work for docker-installed nodes (though bypassing the daemon).

> About the documentation about CRI yes I'll keep adding stuff as I've already done in the past!

Thanks!

> @yujuhong let's leave this open for further discussion

Sure.

**Comment by runcom** [comment]
@yujuhong @Random-Liu @feiskyer this is the description of `StopPodSandbox`:

```

    // StopPodSandbox stops any running process that is part of the sandbox and
    // reclaims network resources (e.g., IP addresses) allocated to the sandbox.
    // If there are any running containers in the sandbox, they must be forcibly
    // terminated.
    // This call is idempotent, and must not return an error if all relevant
    // resources have already been reclaimed. kubelet will call StopPodSandbox
    // at least once before calling RemovePodSandbox. It will also attempt to
    // reclaim resources eagerly, as soon as a sandbox is not needed. Hence,
    // multiple StopPodSandbox calls are expected.
    rpc StopPodSandbox(StopPodSandboxRequest) returns (StopPodSandboxResponse) {}
```
https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/apis/cri/v1alpha1/runtime/api.proto#L29-L30

Can we reword it to state that:

```
Kubelet guarantees at least one call of stop before calling remove
```
related to https://github.com/kubernetes-incubator/cri-o/pull/530#discussion_r122325568

**Comment by feiskyer** [comment]
@runcom Yep, this is guaranteed.

**Comment by sameo** [comment]
@feiskyer It would probably make sense to make that guarantee part of the specification. Getting a remove request should error out if the pod is not stopped, imho.
